Collects, persists and displays the status of a middleware deployment.

josejulio opened the pull request:
<img width="1173" alt="listview-status" src="https://cloud.githubusercontent.com/assets/3845764/16599745/6d0896e6-42c9-11e6-9c47-79c023111e36.png">
<img width="953" alt="status-properties" src="https://cloud.githubusercontent.com/assets/3845764/16599746/6d08e092-42c9-11e6-8f73-f05ad7f4902b.png">

Fixes #9250

abonas commented:
please follow this guide on how to link it to the issue it resolves (you are missing a keyword)
https://help.github.com/articles/closing-issues-via-commit-messages/
and please also add a screenshot of what you've added here

abonas commented:
@miq-bot assign @abonas

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager.rb:
```diff
@@ -157,6 +157,10 @@ def self.default_blacklisted_event_names
     end
 
+    def build_metric_id(type, feed, resource, metric_id)
+      "#{type}I~R~[#{feed}/#{resource.id}]~#{type}T~#{metric_id}"
```
type appears twice - is this ok?

josejulio replied:
Yes, it seems to work that way. Took it from [Hawkular source](https://github.com/hawkular/hawkular-dist-old/blob/master/console/src/main/scripts/plugins/metrics/ts/services/metricsService.ts#L112-L114).

abonas commented on app/helpers/middleware_deployment_helper/textual_summary.rb:
```diff
@@ -23,4 +23,8 @@ def textual_name
   end
+
+  def textual_enabled
+    {:label => 'Enabled/Disabled', :value => @record.enabled ? 'Enabled' : 'Disabled'}
```
iirc the label should be with translation - see other similar files for examples. it has something to do with _(

josejulio replied:
Right. Is like that in the other files. I'll fix this.

josejulio replied:
Should literal value also be translated ?

abonas replied:
yes, given that those are values that are "easy to anticipate"

abonas commented on spec/models/manageiq/providers/hawkular/middleware_manager/refresh_parser_spec.rb:
```diff
@@ -2,25 +2,22 @@
-                       :authentications => [auth],
-                       :zone            => zone)
+    @ems_hawkular = FactoryGirl.create(:ems_hawkular,
+                                       :hostname        => 'hservices.torii.gva.redhat.com',
```
it's better not to have those real server names in the code. this is a mock anyway...

josejulio replied:
I copied from here: https://github.com/josejulio/manageiq/blob/d4af06419d5e5538114f10258b87c589ad3724d1/spec/models/manageiq/providers/hawkular/middleware_manager/refresher_spec.rb#L6
Though it could be useful when re-recording so I used the same. But I'm on the process of removing the changes on this file because I'm moving the logic a bit.

abonas commented on spec/models/manageiq/providers/hawkular/middleware_manager/refresh_parser_spec.rb:
```diff
@@ -2,25 +2,22 @@
-                       :port            => 8080,
-                       :authentications => [auth],
-                       :zone            => zone)
+    @ems_hawkular = FactoryGirl.create(:ems_hawkular,
```
why is this refactoring part of this patch?

abonas commented on spec/models/manageiq/providers/hawkular/middleware_manager/refresh_parser_spec.rb:
```diff
@@ -95,4 +92,29 @@
+
+  describe 'parse_deployment' do
+    it 'handles data and fetches the enabled metric' do
+      # parse_deployment(server, deployment)
```
is this comment needed?

josejulio replied:
Is just a reminder of what I'm testing there, I could remove it, below is clear what I'm testing.

abonas commented on spec/vcr_cassettes/manageiq/providers/hawkular/middleware_manager/refresh_parser.yml:
```diff
@@ -0,0 +1,48 @@
+---
```
Why is this cassette needed?

josejulio replied:
I had the logic of fetching the deployment availability on the parse_deployment. This won't be needed as I'm moving this. (I'll delete this cassette)

abonas commented:
@ammendonca please also review

josejulio commented:
Moved the code to be ready for [HWKMETRICS-393](https://issues.jboss.org/browse/HWKMETRICS-393). Not it does fetch all the availabilities on the fetch_availability method (still one per request, but ready to replace).

ammendonca commented:
I have tested this locally and it seems there's an issue with fetching the availability, which makes the refresh fail and thus no servers are added.

![image](https://cloud.githubusercontent.com/assets/1737954/16418257/ed24de04-3d40-11e6-8aa1-63d697323af4.png)

Confirmed from wireshark that it is on availability endpoint:

![image](https://cloud.githubusercontent.com/assets/1737954/16418333/4171b324-3d41-11e6-98c6-8963ea085ce9.png)

I'm using Hawkular from master.

abonas commented:
@ammendonca is this already working for you or the error still occurs?

ammendonca commented:
Yes, I am still facing the issue, and it seems to be this issue: https://github.com/hawkular/hawkular-client-ruby/issues/69

After investigating, the header was forbidden when metrics were integrated with hawkular-accounts, but it is not integrated anymore (https://issues.jboss.org/browse/HWKMETRICS-399) and the header is required again in latest version.

I've tested with latest hawkular-services (which includes metrics 0.16.0.Final+) and with it, it works fine.

Other than that, would just fix the coherency in deployments list view, the row is called "Enabled/Disabled" and the values are "True" or "False". Would call the row "Status" (also would change it in deployment detail view) and show the value as "Enabled" and "Disabled" (or "Stopped", think that's what we used in Hawkular).

josejulio commented:
@ammendonca I'm having problems changing the display value on the list view. It doesn't seem to support something like 'app/helpers/middleware_deployment_helper/textual_summary.rb', do you know of way to achieve this?

I could change the database to use a string instead of a boolean and store 'enabled'/'disabled' (or 'stopped'). WDYT ?

// cc @abonas

ammendonca commented:
I think it would be good to change it to string, since deployment status is actually a string as well and can be "OK" (deployed and enabled), "STOPPED" (deployed but not enabled) or "FAILED" (failed to deploy due to dependencies or other runtime error).

pilhuhn commented:
@ammendonca Hawkular-master (alpha11) is no longer supported. All work happens against hawkular-services now.

abonas commented:
@josejulio @ammendonca if indeed the status can hold more than 2 values, then I think it would be more correct to convert it to string, otherwise we won't be able to reflect the status correctly in manageiq.

josejulio commented:
Ok, I'll change it to string and rename it to 'status' instead of 'enabled'

josejulio commented:
Did the following changes:
-  Renamed enabled to status and made it a string
-  Updated to use hawkular-client 2.2.0 (fetching all availabilities on one call)
-  Updated screenshots

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb:
```diff
@@ -94,6 +95,18 @@ def fetch_server_entities
       end
 
+      def fetch_availability
+        metrics_ids = {}
```
should this be metrics_ids or metric_id?

josejulio replied:
Its a hash that maps an already fetched resource with its availability id. (In this case there is Deployments, but could be anything else to fetch them in one call). 
I couldn't think of a good name, maybe `resources_by_availability_id` ? WDYT ?

abonas replied:
I'm ok with all the names, was just wondering about the plural part repeated in both "metrics" and "ids". it's really a nit.

josejulio replied:
oh, I'll replace it to metric_ids then.

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb:
```diff
@@ -94,6 +95,18 @@ def fetch_server_entities
+        end
+        availabilities = @ems.metrics_client.avail.raw_data(metrics_ids.keys, :limit => 1)
+        availabilities.each do |availability|
+          metrics_ids[availability['id']][:status] = process_availability(availability['data'].first)
```
does the id arrive as 'id' or :id?

josejulio replied:
metrics_ids contains already parsed data. availability is coming directly from the server. That's the reason it uses `id` instead of `:id`

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb:
```diff
@@ -111,6 +124,17 @@ def process_server_entity(server, entity)
 
+      def process_availability(availability)
+        case
+        when availability.nil?, availability['value'] == 'unknown'
```
is it worth to check here perhaps with downcase? in case the value arrives as Uknown or UKNOWN

josejulio replied:
It *should* always arrive downcase[1]. But then, is only one call and might be worth doing that.

I read somewhere that "Running .empty? on something that is nil will throw a NoMethodError", I could test both if that's the case. "when availability.nil?, availability.empty?,..."

[1] https://github.com/hawkular/hawkular-metrics/blob/master/core/metrics-model/src/main/java/org/hawkular/metrics/model/AvailabilityType.java#L28-L34

abonas replied:
good point. then perhaps `blank?` would be a better fit

abonas commented on app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb:
```diff
@@ -111,6 +124,17 @@ def process_server_entity(server, entity)
+      def process_availability(availability)
+        case
+        when availability.nil?, availability['value'] == 'unknown'
+          _('Unknown')
```
iiic, you are using here translation/i18n, but this is a backend parser. 
translation typically belongs to frontend to allow flexibility per several different languages for different users.
if the string is translated here once in backend, that ability would be lost.

josejulio replied:
I agree with you on that part. Not translating on backend would benefit the clients for using whatever language they want or even change the language at runtime.
If that's how is done in MiQ I'll remove the translation call.

abonas commented:
@miq-bot assign @ammendonca

abonas commented:
@ammendonca please give this a review since the code was updated. thanks!

ammendonca commented:
LGTM

abonas commented:
@miq-bot assign @chessbyte
